source/common/router/router.cc

```diff
@@ -1125,7 +1125,14 @@ void Filter::onUpstream100ContinueHeaders(Http::ResponseHeaderMapPtr&& headers,
   // the complexity until someone asks for it.
   retry_state_.reset();
 
-  callbacks_->encode100ContinueHeaders(std::move(headers));
+  // We coalesce 100-continue headers here, to prevent encoder filters and HCM from having to worry
+  // about this. This is done in the router filter, rather than UpstreamRequest, since we want to
+  // potentially coalesce across retries and multiple upstream requests in the future, even though
+  // we currently don't support retry after 100.
+  if (!downstream_100_continue_headers_encoded_) {
+    downstream_100_continue_headers_encoded_ = true;
+    callbacks_->encode100ContinueHeaders(std::move(headers));
```

Comment:
Sorry I think my last comment wasn't sufficiently clear.  Is it doable to do this in the HCM?  This gets us retries which is great, but not internal redirects which would be a perk.   I'm OK with that if it's a pain to move, but if it's simple to move it get us one more feature.

Reply:
As discussed offline, going to punt on this one, since it's complicated by the fact that today we have `codec --> upstream request --> router encode filter --> other encode filters --> HCM` on the encode path, and we want to avoid having rando encode filters needing to worry about > 1 `encode100ContinueHeaders()` or `encodeHeaders()`.